File: customer-java-resttemplate/src/test/java/com/prez/UsingRedis.java
package com.prez;

import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.support.TestPropertySourceUtils;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;
import org.testcontainers.utility.DockerImageName;

@ContextConfiguration(initializers = {UsingRedis.Initializer.class})
@Testcontainers
public class UsingRedis {
  public static final int REDIS_PORT = 6379;

  @Container
  public static GenericContainer<?> redis = new GenericContainer<>(DockerImageName.parse("redis:6.0.9"))
      .withExposedPorts(REDIS_PORT)
      .withReuse(true);

  static class Initializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {
    @Override
    public void initialize(ConfigurableApplicationContext applicationContext) {
      redis.start();

      String redisContainerPort = "spring.redis.port=" + redis.getMappedPort(REDIS_PORT);
      TestPropertySourceUtils.addInlinedPropertiesToEnvironment(applicationContext, redisContainerPort);
    }
  }

}

File: customer-kotlin-coroutines/src/test/kotlin/com/prez/UsingRedis.kt
package com.prez

import org.springframework.boot.test.util.TestPropertyValues
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.test.context.ContextConfiguration
import org.testcontainers.containers.GenericContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.utility.DockerImageName

@ContextConfiguration(initializers = [UsingRedis.Initializer::class])
open class UsingRedis {

  companion object {
    const val REDIS_PORT = 6379

    @Container
    var redis: GenericContainer<Nothing> = GenericContainer<Nothing>(DockerImageName.parse("redis:6.2"))
      .withExposedPorts(REDIS_PORT)
  }

  internal class Initializer : ApplicationContextInitializer<ConfigurableApplicationContext> {
    override fun initialize(applicationContext: ConfigurableApplicationContext) {
      redis.withReuse(true) // another difference with Java version
      redis.start()
      TestPropertyValues.of(
        "spring.redis.port=" + redis.getMappedPort(REDIS_PORT)
      ).applyTo(applicationContext.environment)
    }
  }

}
